src/methods.jl

```diff
@@ -10,318 +19,453 @@ derivatives.
-differencing.
-See [`FiniteDifferenceMethod`](@ref) for further details.
+# Returns
+- Estimate of the derivative.
+
+# Examples
+
+```julia-repl
+julia> fdm = central_fdm(5, 1)
+FiniteDifferenceMethod:
+  order of method:       5
+  order of derivative:   1
+  grid:                  [-2, -1, 0, 1, 2]
+  coefficients:          [0.08333333333333333, -0.6666666666666666, 0.0, 0.6666666666666666, -0.08333333333333333]
+
+julia> fdm(sin, 1, 1e-3)
+0.5403023058679624
+
+julia> fdm(sin, 1, 1e-3) - cos(1)  # Check the error.
+-1.7741363933510002e-13
+```
 """
-function Nonstandard(grid::AbstractVector{<:Real}, q::Integer; adapt=0, kwargs...)
-    p = length(grid)
-    _check_p_q(p, q)
-    coefs = _coefs(grid, p, q)
-    hist = History(; adapt=adapt, kwargs...)
-    return Nonstandard{typeof(grid), typeof(coefs)}(Int(p), Int(q), grid, coefs, hist)
+function (m::FiniteDifferenceMethod)(
+    f::Function,
+    x::T,
+    h
+) where T<:AbstractFloat
+    return sum(
+        i -> convert(T, m.coefs[i]) * f(T(x + h * m.grid[i])),
+        eachindex(m.grid)
+    ) / h^m.q
 end
+# Handle arguments that are not floats. Assume that converting to float is desired.
+(m::FiniteDifferenceMethod)(f::Function, x::T, h) where T<:Real = m(f, float(x), h)
 
-# Check the method and derivative orders for consistency
+# Check the method and derivative orders for consistency.
 function _check_p_q(p::Integer, q::Integer)
-    q >= 0 || throw(ArgumentError("order of derivative must be nonnegative"))
+    q >= 0 || throw(ArgumentError("order of derivative must be non-negative"))
     q < p || throw(ArgumentError("order of the method must be strictly greater than that " *
                                  "of the derivative"))
-    # Check whether the method can be computed. We require the factorial of the
-    # method order to be computable with regular `Int`s, but `factorial` will overflow
-    # after 20, so 20 is the largest we can allow.
+    # Check whether the method can be computed. We require the factorial of the method order
+    # to be computable with regular `Int`s, but `factorial` will after 20, so 20 is the
+    # largest we can allow.
     p > 20 && throw(ArgumentError("order of the method is too large to be computed"))
     return
 end
 
-# Compute coefficients for the method
+const _COEFFS_CACHE = Dict{Tuple{AbstractVector{<:Real}, Integer}, Vector{Float64}}()
 
-const _COEFFS_CACHE = Dict{Tuple{AbstractVector{<:Real}, Integer, Integer}, Vector{Float64}}()
-function _coefs(grid::AbstractVector{<:Real}, p::Integer, q::Integer)
-    return get!(_COEFFS_CACHE, (grid, p, q)) do
-        # For high precision on the \ we use Rational, and to prevent overflows we use Int128
-        # At the end we go to Float64 for fast floating point math (rather than rational math)
+# Compute coefficients for the method and cache the result.
+function _coefs(grid::AbstractVector{<:Real}, q::Integer)
+    return get!(_COEFFS_CACHE, (grid, q)) do
+        p = length(grid)
```

Comment:
That is a nice improvement to not have to remember `p`

Reply:
Yes, it seemed redundant to carry around both.